cmake/Modules/Packages/USER-PLUMED.cmake

```diff
@@ -66,31 +69,34 @@ if(PKG_USER-PLUMED)
       BUILD_BYPRODUCTS ${PLUMED_BUILD_BYPRODUCTS} 
     )
     ExternalProject_get_property(plumed_build INSTALL_DIR)
-    set(PLUMED_INSTALL_DIR ${INSTALL_DIR})
-    list(APPEND LAMMPS_DEPS plumed_build)
+    add_library(LAMMPS::PLUMED UNKNOWN IMPORTED)
```

Comment:
@GiovanniBussi please have a look at these changes, it would be nice if a future version of plumed could just export targets, so that we can get rid of all that custom craft here.

Reply:
@junghans I apologise but I am not really familiar with cmake syntax so I can't judge...

Can you be more precise about which change you suggest we do uptstream? Currently, we provide a cmake include file that contains the following variables:
```
PLUMED_LOAD # all the objects or libraries that should be linked (including -L flags)
PLUMED_DEPENDENCIES # all files that, if changed, require a recompilation - not used in many places though
```
If you think it is useful we can provide separate variables (e.g. `PLUMED_OBJECTS`, `PLUMED_LDFLAGS` etc), but I would keep `PLUMED_LOAD` as is for backward compatibility.

I also add @gtribello who actually made most of the work for integrating PLUMED in LAMMPS. Thanks!

Reply:
@GiovanniBussi  on a second thought as you guys are using autotools it might be pretty hard to create exported targets for cmake. However, if you provide 3 three pkg-config files (instead of one) for static, shared, runtime, e.g. `plumed_shared.pc`  that include all the c++ defines as well we could do something like:
```
find_package(PkgConfig REQUIRED)
pkg_check_modules(PLUMED REQUIRED plumed_${PLUMED_MODE}
target_link_libraries(lammps PRIVATE PkgConfig::PLUMED)
```
in cmake and
```
plumed_SYSINC=`pkg-config --cflags plumed_${PLUMED_MODE})
plumed_SYSLIB=`pkg-config --libs plumed_${PLUMED_MODE})`
```
in the Makefiles and hence remove all the hard-coded `__PLUMED` defines.

Reply:
I see. If you are happy with pkgfiles, notice that we already provide two of them:
1. `plumed.pc` that in principle can be used for both "shared" and "static" mode. In particular, for static binding one should link `Libs.private` as well. When looking at pkgconfig documentation some time ago I understood this is the way to distinguish static and shared library linking (static libraries need dependencies to be specified as well).
2. `plumedWrapper.pc` that can be used to link plumed in "runtime" mode.

I paste below the two files as they are generated on my Mac when I install plumed with MacPorts so that you can easily see what they contain.

Do you think this is sufficient/correct?

Many thanks!

Giovanni

plumed.pc:
```
prefix=/opt/local
exec_prefix=/opt/local
includedir=/opt/local/include
libdir=/opt/local/lib

Name: plumed
Description: Plumed library
URL: http://www.plumed.org
Version: 2.6.0
Cflags: -I${includedir}
Libs: -L${libdir} -lplumed
Libs.private: -ldl  -ldl  -L/opt/local/lib -Wl,-headerpad_max_install_names -lxdrfile -lz -lgsl -stdlib=libc++ -lvecLibFort -Wl,-syslibroot,/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk -arch x86_64
Conflicts:
```

plumedWrapper.pc:
```
prefix=/opt/local
exec_prefix=/opt/local
includedir=/opt/local/include
libdir=/opt/local/lib

Name: plumedWrapper
Description: Plumed library  (wrapper-only version, set PLUMED_KERNEL at runtime)
URL: http://www.plumed.org
Version: 2.6.0
Cflags: -I${includedir}
Libs: -L${libdir} -lplumedWrapper -ldl
Libs.private:
Conflicts:
```

Reply:
These two miss the magic `-D__PLUMED`  flags in the `Cflags:` line.

Reply:
Is this required/recommended? We currently do not use that macro anywhere in PLUMED source code (`git grep "__PLUMED[^_]"` returns nothing in plumed tree). Or is this the standard way to say a package is available?

Reply:
Hmm, that is interesting, we also seem to only use `__PLUMED_DEFAULT_KERNEL` in one place in the lammps code:
```
git grep __PLUMED
...
src/USER-PLUMED/fix_plumed.cpp:#if defined(__PLUMED_DEFAULT_KERNEL)
src/USER-PLUMED/fix_plumed.cpp:static char plumed_default_kernel[] = "PLUMED_KERNEL=" PLUMED_QUOTE(__PLUMED_DEFAULT_KERNEL);
src/USER-PLUMED/fix_plumed.cpp:#if defined(__PLUMED_DEFAULT_KERNEL)
```

@akohlmey ?

Reply:
A quick comment about `__PLUMED_DEFAULT_KERNEL`. This macro is not necessary if you link plumed in one of the three officially supported modes (static/shared/runtime). It is used internally by plumed when compiling the `libplumedWrapper.a` library (the one linked to MD codes in runtime mode) and could be used in MD codes that embed the plumed wrappers rather then linking them (this is the case for AmberMD). It provides a default path for `dlopen` so that you can run an MD code loading PLUMED without the need to specify the environment variable `PLUMED_KERNEL`.